### ISSUE TITLE
Unready Kubernetes pods reachable via loadbalancer on Istio v1.14.1

### ISSUE THREAD

**Issue by schmigware** [istio/istio#39825]
### Bug Description

For a simple configuration:

- Some pod with configured readiness probe
- Aforementioned pod is load-balanced by a ClusterIP service
- Istio installed

I expect that when a load-balanced pod is not ready, it should not be reachable via the load-balancing service.

With Istio 1.14.1, if I configure:
- Some pod(s) with ready status FALSE
- ClusterIP service (load-balancer)
-  A pair of pods from which to issue curl requests, one with an envoy sidecar and one without

Then:

- The pod with an envoy sidecar can successfully e.g. curl an unready endpoint via the load balancer. 
- The pod without any envoy sidecar cannot.

For earlier Istio versions - specifically 1.13.2 for instance - I observe:
- The pod with an envoy sidecar cannot successfully e.g. curl an endpoint, returning status **no healthy upstream**
- The pod without any envoy sidecar cannot reach any loadbalanced pod.

The behaviour exhibited by version 1.13.2 is the behaviour I expect. 

Please clarify, many thanks.
 

### Version

```prose
client version: 1.14.1
control plane version: 1.14.1
data plane version: 1.13.2 (2 proxies), 1.14.1 (2 proxies)

Client Version: v1.24.2
Kustomize Version: v4.5.4
Server Version: v1.23.8+k3s1

v3.9.0+g7ceeda6
```

**Comment by zirain** [comment]
may related to https://github.com/istio/istio/pull/37122, cc @ramaraochavali

**Comment by ramaraochavali** [comment]
Can you please share /config_dump?include_eds output on the proxy and share the service name?

**Comment by schmigware** [comment]
[config_dump.log](https://github.com/istio/istio/files/9063180/config_dump.log)

**Comment by ramaraochavali** [comment]
It shows both the endpoints as `UNHEALTHY` - which is correct. What is the curl output you are getting?

```
  "endpoints": [
       {
        "locality": {},
        "lb_endpoints": [
         {
          "endpoint": {
           "address": {
            "socket_address": {
             "address": "10.42.0.250",
             "port_value": 8080
            }
           },
           "health_check_config": {}
          },
          "health_status": "UNHEALTHY",
          "metadata": {
           "filter_metadata": {
            "istio": {
             "workload": "aggregator;dcm;aggregator;latest;Kubernetes"
            },
            "envoy.transport_socket_match": {
             "tlsMode": "istio"
            }
           }
          },
          "load_balancing_weight": 1
         },
         {
          "endpoint": {
           "address": {
            "socket_address": {
             "address": "10.42.0.251",
             "port_value": 8080
            }
           },
           "health_check_config": {}
          },
          "health_status": "UNHEALTHY",
          "metadata": {
           "filter_metadata": {
            "istio": {
             "workload": "aggregator;dcm;aggregator;latest;Kubernetes"
            },
            "envoy.transport_socket_match": {
             "tlsMode": "istio"
            }
           }
          },
          "load_balancing_weight": 1
         }
        ],
        "load_balancing_weight": 2
       }
      ],
      "policy": {
       "overprovisioning_factor": 140
      }
     }
    },
```

**Comment by schmigware** [comment]
I'm curling a prometheus /metrics endpoint:

shell  ~  **curl aggregator-lb.dcm.svc.cluster.local:8080/metrics** 
# HELP go_gc_duration_seconds A summary of the pause duration of garbage collection cycles.
# TYPE go_gc_duration_seconds summary
go_gc_duration_seconds{quantile="0"} 2.7364e-05
go_gc_duration_seconds{quantile="0.25"} 3.7239e-05
go_gc_duration_seconds{quantile="0.5"} 4.834e-05
go_gc_duration_seconds{quantile="0.75"} 6.1627e-05
go_gc_duration_seconds{quantile="1"} 0.000164292
go_gc_duration_seconds_sum 0.004197202
go_gc_duration_seconds_count 75

**Comment by schmigware** [comment]
Same command with 1.13.2 yields **no healthy upstream**

**Comment by ramaraochavali** [comment]
> Same command with 1.13.2 yields no healthy upstream

Yes. Because we were not sending un healthy endpoints. Envoy should not use it for load balancing.

**Comment by ramaraochavali** [comment]
> As I expect. Has something changed in 1.14.1? The end result is different for an otherwise identical configuration.

Yes. In 1.14.1 Istio is sending unhealthy endpoints to Envoy to support slow start mode. You can set `PILOT_SEND_UNHEALTHY_ENDPOINTS` to false to get back the old behaviour.

I suspect this has to do with panic threshold.  Is this just to test or did you face any issue because of this?

**Comment by schmigware** [comment]
Should this change in behaviour across Istio versions not be called out front-and-centre in documentation? Effectively, Istio 1.14.x - out of the box - appears configured to ignore Kubernetes readiness semantics.

**Comment by ramaraochavali** [comment]
It is called out in the release notes https://istio.io/latest/news/releases/1.14.x/announcing-1.14/change-notes/#traffic-management

**Comment by schmigware** [comment]
You are expecting the reader to infer that Kubernetes readiness semantics are not going to apply out of the box anymore once an Istio upgrade has been perforrmed - from the first note? Surely there should be a warning of some sort. This change can have serious implications for production deployments.

Suggest updating the documentation to read e.g.:

_Added support for sending unready endpoints to Envoy. This will be useful when slow start mode in Envoy is enabled. This can be disabled by setting PILOT_SEND_UNHEALTHY_ENDPOINTS to false. **WARNING**: As a result of this change, Istio's default behaviour is to route traffic to unhealthy pods._

**Comment by ramaraochavali** [comment]
> WARNING: As a result of this change, Istio's default behaviour is to route traffic to unready endpoints.

While I agree it should have been more clear, this is not entirely correct. By default it wont route it to unready endpoints. Please read the https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/upstream/load_balancing/panic_threshold.html?highlight=panic%20threshold. If all the endpoints are unhealthy, then the load balancer will get in to panic mode and routes to unhealthy endpoints (Sure earlier unready ones were not sent, but it can happen to endpoints that are eady but unhealthy as well when it gets in to that mode).

The flag can not be set to false because it is needed for https://github.com/istio/istio/issues/21228.

I will update the doc to be more clear.

**Comment by howardjohn** [comment]
@ramaraochavali this is IMO a super super severe bug and we should revert the change or find a way to make it not trigger the healthy panic threshold. Kubernetes defines a clear contract not to send traffic to unready endpoints and we are violating that

**Comment by ramaraochavali** [comment]
I was thinking about that as well - I think the best path forward here is to default it false and when it is enabled, disable panic threshold - while we loose some benefits of panic threshold, but it is better.

**Comment by hzxuzhonghu** [comment]
I would vote option 2, because option 1 equals to removing panic threshold feature, which is a breaking change.  Users of istio have run stably more or less for several years under panic threshold functionality. While the feature flag is on in 1.14, I suspect no production user yet.

**Comment by hzxuzhonghu** [comment]
@ramaraochavali Does setting panic threshold to 0 has any influence to priority failover?  Or any other severe influence on system stability?

**Comment by ramaraochavali** [comment]
Both are breaking changes

> Users of istio have run stably more or less for several years under panic threshold functionality.

Most of the users specify/use outlier detection and when outlier detection is provided without min_health_percent , we already disable it https://github.com/istio/istio/blob/master/pilot/pkg/networking/core/v1alpha3/cluster.go#L663

> Does setting panic threshold to 0 has any influence to priority failover? Or any other severe influence on system stability?

It does not have influence on priority failover but priority load balancing is impacted as described https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/upstream/load_balancing/panic_threshold.html?highlight=panic%20threshold

> While the feature flag is on in 1.14, I suspect no production user yet.

Fair point.  My motivation is to handle cases documented https://github.com/istio/istio/issues/18367 (specifically comments from here https://github.com/istio/istio/issues/18367#issuecomment-550559527 failover does not work properly) - This is a long standing feature request.

Having said that, I am definitely open to option #2.

**Comment by ramaraochavali** [comment]
So  I did the following in the PR which I think is a middle ground
- Defaulted `PILOT_SEND_UNHEALTHY_ENDPOINTS` to false
- When enabled, disabled panic threshold. Updated its docs.
- Will send another PR to update update https://github.com/istio/api/blob/master/networking/v1alpha3/destination_rule.proto#L619 to indicate PILOT_SEND_UNHEALTHY_ENDPOINTS should be enabled to handle readiness blips not to cause spurious warmups.
